bsp/src/org/jetbrains/bsp/project/test/BspTestRunProfileState.scala

```diff
@@ -0,0 +1,145 @@
+import com.intellij.execution.testframework.sm.runner.SMTRunnerConsoleProperties
+import com.intellij.execution.{DefaultExecutionResult, ExecutionResult, Executor}
+import com.intellij.openapi.module.ModuleManager
+import com.intellij.openapi.project.Project
+import jetbrains.buildServer.messages.serviceMessages._
+import org.jetbrains.bsp.data.BspMetadata
+import org.jetbrains.bsp.protocol.BspCommunicationComponent
+import org.jetbrains.bsp.protocol.BspNotifications.{BspNotification, TaskFinish, TaskStart}
+import org.jetbrains.bsp.protocol.session.BspSession.BspServer
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext.Implicits._
+
+class BspTestRunProfileState(val project: Project, val rc: BspTestRunConfiguration, val ex: Executor) extends RunProfileState {
+
+  val gson = new Gson()
+
+  class MProcHandler extends ProcessHandler {
+    override def destroyProcessImpl(): Unit = ()
+
+    override def detachProcessImpl(): Unit = {}
+
+    override def detachIsDefault(): Boolean = false
+
+    override def getProcessInput: OutputStream = null
+
+    def shutdown(): Unit = {
+      super.notifyProcessTerminated(0)
+    }
+  }
+
+  private def targets(): List[URI] = {
+    ModuleManager.getInstance(project).getModules.toList
+      .flatMap(BspMetadata.get(project, _))
+      .flatMap(x => x.targetIds.asScala.toList)
+  }
+
+  private def testRequest(server: BspServer): CompletableFuture[TestResult] = {
+    val targetIds = targets().map(uri => new BuildTargetIdentifier(uri.toString))
+    val params = new TestParams(targetIds.toList.asJava)
+    params.setOriginId(UUID.randomUUID().toString)
+    server.buildTargetTest(params)
+  }
+
+
+  def printProc(str: MessageWithAttributes)(implicit proc: ProcessHandler): Unit = {
+    proc.notifyTextAvailable(str + System.lineSeparator(), ProcessOutputType.STDOUT)
+  }
+
+  class ServiceMsg(val name: String, val map: Map[String, String]) extends MessageWithAttributes(name, map.asJava) {
+    def this(n: String) = this(n, Map())
+  }
+
+  case class NestedBspMsg(nested: AnyRef, msg: AnyRef)
+
+  type HasNested = {
+    def getDataKind: String
+    def getData: Object
+  }
+
+  private def extractNestedMessage(bspMessage: AnyRef): AnyRef = bspMessage match {
+    case x: HasNested@unchecked =>
+      val kind = x.getDataKind
+      val rawNested = x.getData
+      rawNested match {
+        case d: JsonObject => kind match {
+          case TEST_FINISH => gson.fromJson(d, classOf[TestFinish])
+          case TEST_TASK => gson.fromJson(d, classOf[TestTask])
+          case TEST_REPORT => gson.fromJson(d, classOf[TestReport])
+          case TEST_START => gson.fromJson(d, classOf[TestStart])
+          case _ => rawNested
+        }
+        case _ => rawNested
+      }
+    case x => x
+  }
+
+  class BspTestSession {
+    val startTime: mutable.Map[String, Long] = mutable.Map()
+  }
+
+  // TODO handle standard output of test cases
```

Comment:
Handling log messages is a little difficult. There is no information about which test did cause a `LogMessage`, however tree view requires to indicate that explicitly. We can store current test in the `BspTestSession` and pass that to tree view, but I am not sure if that's a good idea. For example if tests are being run concurrently it might brake, however in that case no one knows which log message comes from which test, unless there's explicit synchronization

Reply:
`LogMessageParams` does have a `task` field with the task id of the originating task. This *should* be the id of the specific test which is being executed. If it is not available, it's ok if goes into the test run root output